### PR TITLE
New statistics registry

### PR DESCRIPTION
**Pull request by nafur**
This PR adds the next part of the new statistics setup: the registry.
The new statistics registry owns the actual data and only issues proxy objects that can be used to modify the internally stored data.
Once we replace the old statistics setup, the files should be renamed from `statistics_reg.*` to `statistics_registry.*`.

It depends on #6178 being merged.

### CODE REVIEW

File: src/util/CMakeLists.txt [hunk 1]
```diff
@@ -56,8 +56,10 @@ libcvc4_add_sources(
   sexpr.h
   smt2_quote_string.cpp
   smt2_quote_string.h
-  statistics.cpp
-  statistics.h
+  statistics_public.cpp
+  statistics_public.h
+  statistics_reg.cpp
+  statistics_reg.h
```
Comment (mpreiner) [inline src/util/CMakeLists.txt:62]: Will these files be renamed when we remove the old statistics registry?

File: src/util/statistics_reg.h [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+ * unexpected (though not undefined) behaviour for others. For a reference
+ * statistic, for example, the internal data will simply point to the object
+ * registered last.
+ * Note that the type of the re-registered statistic must always match
+ * the type of the previously registered statistic with the same name.
+ *
+ * We generally distinguish between public (non-expert) and private (expert)
+ * statistics. By default, `--stats` only shows public statistics. Private
+ * ones are printed as well if `--all-statistics` is set.
+ * All registration method have a trailing argument `expert`, defaulting to
```
Comment (mpreiner) [inline src/util/statistics_reg.h:56]:
```suggestion
 * All registration methods have a trailing argument `expert`, defaulting to
```

File: src/util/CMakeLists.txt (hunk 1 above)
Comment (nafur) [reply src/util/CMakeLists.txt:62]: Yes.

File: src/util/statistics_reg.cpp [hunk 3]
```diff
@@ -54,25 +71,58 @@ StatisticBaseValue* StatisticsRegistry::get(const std::string& name) const
...
-    os << s.first << " = ";
-    s.second->print(os);
-    os << std::endl;
+  if constexpr (Configuration::isStatisticsBuild()) {
+    for (const auto& s : d_stats)
+    {
+      if (!options::statisticsExpert() && s.second->d_expert) continue;
+      if (!options::statisticsUnset() && !s.second->hasValue()) continue;
+      os << s.first << " = " << *s.second << std::endl;
+    }
```
Comment (mpreiner) [inline src/util/statistics_reg.cpp:82]: clang-format

File: src/util/statistics_reg.h [hunk 4]
```diff
@@ -193,6 +211,8 @@ class StatisticsRegistry
    * registered for.
    */
   std::map<std::string, std::unique_ptr<StatisticBaseValue>> d_stats;
+
+  std::unique_ptr<Snapshot> d_lastSnapshot;
```
Comment (mpreiner) [inline src/util/statistics_reg.h:215]: Add documentatino for `d_lastSnapshot`

**Review by mpreiner — APPROVED** [review]

**Pull request merged**